### ISSUE TITLE
publish tools and datasets to enable external researchers to evaluate proposal w.r.t. privacy

### ISSUE THREAD

**Issue by aberke** [patcg-individual-drafts/topics#49]
It's great that this proposal is incorporating public feedback. 

It would be even better if this proposal was published with a set of tools and datasets for external researchers and the web community to better evaluate the proposal with empirical tests.

For example, a colleague and I recently did a post-mortem analysis of FLoC:
https://arxiv.org/pdf/2201.13402.pdf
Our analysis required us to re-implement FloC and to leverage a proprietary dataset of browsing histories. These hurdles make analysis by and for the public inaccessible to many external researchers and community members. 

It would be helpful for the Chrome developers to publish tools, example datasets, and code so that their proposals can be more easily interrogated by researchers. Will such be made available?

**Comment by jkarlin** [comment]
Very cool about your paper. We do not have a public dataset of browsing histories to suggest. Our code (including FLoC) is all open-source, so you can replicate our algorithms exactly. Note that there is no landed code for Topics yet, so I can't offer anything to look at right now.

What sorts of tools would be of interest?

**Comment by aberke** [comment]
We appreciate the positive reception!

> What sorts of tools would be of interest?

The Topics designers/engineers and others posting issues likely have many more ideas. 
Here is a first example of to open up a larger discussion: a sample dataset for analysis.

Our FLoC empirical analysis used a proprietary dataset of browsing histories and user demographics from comScore. They and other data brokers sell such datasets at costs that may be high for many researchers; it would seemingly be possible for an organization like Google to make such a dataset more available (and privacy safe) for analysis.

> We do not have a public dataset of browsing histories to suggest.

The FLoC OT implementation used browsing histories from Chrome users to compute initial cohorts and determine which were sensitive and should be dropped ([the t-closeness analysis](https://docs.google.com/a/google.com/viewer?a=v&pid=sites&srcid=Y2hyb21pdW0ub3JnfGRldnxneDo1Mzg4MjYzOWI2MzU2NDgw)) -- it would seem such datasets may then be available. 

It might be useful to release an anonymized sample of browsing histories / domain visits network graph. Such a dataset might relabel the domains with arbitrary identifiers.
To evaluate privacy implications of the Topics API without evaluating the topics themselves, the domains could be labeled with their topics but where the topics are relabeled with identifiers that lack semantic meaning.
Like with our use of the comScore browsing histories dataset, associating user demographics with the browsing histories would help researchers analyze whether the Topics API risks exposing sensitive user demographic information.
Such a dataset could also help analyze the privacy risks of user re-identification, and other questions raised in the project issues.

> ​​Our code (including FLoC) is all open-source, so you can replicate our algorithms exactly. Note that there is no landed code for Topics yet

The FLoC code was made open-source via its Go implementation within the Chromium browser, but  was not directly usable for many analysis tasks ([link](https://source.chromium.org/chromium/chromium/src/+/master:components/federated_learning/), [link](https://source.chromium.org/chromium/chromium/src/+/master:chrome/browser/federated_learning/)). 

*It would have been helpful to release a standalone API that could take a browsing histories dataset and compute cohorts.*

Using the implementation directly built into the Chrome browser meant using the cohorts specifically calculated for the OT with Chrome's huge user base and large k.
This was not suitable for external analysis like ours with a dataset which had a fraction of the number of OT trial users and would have resulted in sparsely populated cohorts for k=2000. For this reason we implemented the FLoC API to allow us to compute cohorts for varying numbers of users and levels of k. 
However, analysis would be more easy to distribute among researchers and verify by the Chrome team if such a standalone API were developed and documented by the FLoC designers.

**Comment by aberke** [comment]
We worked on this analysis with Dr. Latanya Sweeney and would be happy to figure out how we can further work with you on this.

**Comment by jkarlin** [comment]
> It might be useful to release an anonymized sample of browsing histories / domain visits network graph. Such a dataset might relabel the domains with arbitrary identifiers.

This isn't something that I can offer. I agree that it would be nice, but such a dataset would need to be considered very carefully and it's not currently on our radar to do so.

> The FLoC code was made open-source via its Go implementation within the Chromium browser, but was not directly usable for many analysis tasks ([link](https://source.chromium.org/chromium/chromium/src/+/master:components/federated_learning/), [link](https://source.chromium.org/chromium/chromium/src/+/master:chrome/browser/federated_learning/)).

Note that that's in C++ not Go. While it would be nice to have a separate library to make it easier to study this, that's also not high on our current list of priorities. Given that the implementation code is quite small, we feel that it is straight-forward to reproduce. We do anticipate having documentation for how to find and use the upcoming model that maps hostnames to topics however.

**Comment by aberke** [comment]
> Note that that's in C++ not Go. While it would be nice to have a separate library to make it easier to study this, that's also not high on our current list of priorities. Given that the implementation code is quite small, we feel that it is straight-forward to reproduce. We do anticipate having documentation for how to find and use the upcoming model that maps hostnames to topics however.

Apologies for crossing the wires here. Yes -- the browser implementation was in C++ and our analysis built on top of an open source [Go library developed by shigeki](https://github.com/shigeki/floc_simulator) that was meant to reimplement FLoC, and [discussed with michaelkleber](https://github.com/WICG/floc/issues/95).
A hurdle here was that these implementations used a precomputed file mapping simhash to cohort based on collected Chrome browsing histories, so development still needed to be done to make a new library that could support other sets of browsing histories or levels of k for k-anonymity.

The bottom section of the Topics README frames this as a proposal seeking feedback. 
Our point is that it would be easier for the community to evaluate and respond to the proposal if more tools were made available. 
We understand this might be difficult given privacy concerns, and not yet possible while the proposal is still being better defined. However, we hope such tools can be rolled out in parallel to the OT.

**Comment by jkarlin** [comment]
Yes, it's a proposal seeking feedback. We've provided the  algorithms, open-sourced the code, and provided theoretical analysis. However, real user browsing data is not something we're interested in providing. We're considering realistic synthetic data, but I'm not sure that would be satisfying for your case.
